Registry reports failure to trim if trim count is 0

### Problem description
Right now running connectors we sometimes see:

```
Failed to DELETE with sourceTagToPreserve 3ebb269f-eb88-45b9-b0f3-c8bd1d8409a1 and sourceId environment 1 retries left. Status code: 400, body:
{
  "message": "An error occurred while processing your request."
}
(node:1) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Status code: 400, body:
{
  "message": "An error occurred while processing your request."
}
```

In the registry-api this looks like:

```
[ERROR] [08/30/2018 04:46:15.042] [registry-api-akka.actor.default-dispatcher-24] [RecordsService(akka://registry-api)] An error occurred while trimming
scala.MatchError: Success(0) (of class scala.util.Success)
	at au.csiro.data61.magda.registry.DefaultRecordPersistence$.trimRecordsBySource(RecordPersistence.scala:457)
	at au.csiro.data61.magda.registry.RecordsService$$anonfun$trimBySourceTag$1$$anonfun$apply$23$$anonfun$apply$24$$anonfun$apply$25$$anonfun$10$$anonfun$11.apply(RecordsService.scala:188)
	at au.csiro.data61.magda.registry.RecordsService$$anonfun$trimBySourceTag$1$$anonfun$apply$23$$anonfun$apply$24$$anonfun$apply$25$$anonfun$10$$anonfun$11.apply(RecordsService.scala:187)
	at scalikejdbc.DBConnection$$anonfun$3.apply(DBConnection.scala:326)
	at scalikejdbc.DBConnection$class.scalikejdbc$DBConnection$$rollbackIfThrowable(DBConnection.scala:290)
	at scalikejdbc.DBConnection$class.localTx(DBConnection.scala:319)
	at scalikejdbc.DB.localTx(DB.scala:60)
	at scalikejdbc.DB$.localTx(DB.scala:263)
	at au.csiro.data61.magda.registry.RecordsService$$anonfun$trimBySourceTag$1$$anonfun$apply$23$$anonfun$apply$24$$anonfun$apply$25$$anonfun$10.apply(RecordsService.scala:187)
	at au.csiro.data61.magda.registry.RecordsService$$anonfun$trimBySourceTag$1$$anonfun$apply$23$$anonfun$apply$24$$anonfun$apply$25$$anonfun$10.apply(RecordsService.scala:187)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:39)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:415)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

So it appears that if you do a trim but there's no old records to delete, scala fails to match `Success(0)` as an instance of `Success[Long]` and hence rejects it, which resolves as a 400 😕.

On the up side, it does actually do the trim. Just _looks_ like it's failing.

### Problem reproduction steps
Run the environment connector and wait for it to finish.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -20,6 +20,7 @@
 -   Take open data connector license from dataset level to distribution level and add basic black box test
 -   Fix logo vertical alignment and partially hidden issue
 -   Made header padding even
+-   Fixed trim with zero records deleted returning 400
 
 ## 0.0.47
 

--- a/magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala
+++ b/magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala
@@ -442,7 +442,7 @@ object DefaultRecordPersistence extends Protocols with DiffsonProtocol with Reco
     }
 
     val result = recordIds match {
-      case Success(Nil) => Success(0)
+      case Success(Nil) => Success(0l)
       case Success(ids) =>
         ids
           .map(recordId => deleteRecord(session, recordId))

--- a/magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordsService.scala
+++ b/magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordsService.scala
@@ -132,6 +132,7 @@ class RecordsService(config: Config, webHookActor: ActorRef, authClient: AuthApi
             case Success(result)                             => complete(MultipleDeleteResult(result))
             case Failure(timeoutException: TimeoutException) => complete(StatusCodes.Accepted)
             case Failure(exception) =>
+              logger.error(exception, "An error occurred while trimming with sourceTagToPreserve $1 sourceId $2", sourceTagToPreserve, sourceId)
               complete(StatusCodes.BadRequest, BadRequest("An error occurred while processing your request."))
           }
 

--- a/magda-registry-api/src/test/scala/au/csiro/data61/magda/registry/RecordsServiceSpec.scala
+++ b/magda-registry-api/src/test/scala/au/csiro/data61/magda/registry/RecordsServiceSpec.scala
@@ -1690,6 +1690,48 @@ class RecordsServiceSpec extends ApiSpec {
           }
         }
 
+        it("returns HTTP 200 if there's nothing to delete") { param =>
+
+          val junitFile = new java.io.File("../magda-registry-aspects/source.schema.json").getCanonicalFile
+          val commandLineFile = new java.io.File("./magda-registry-aspects/source.schema.json").getCanonicalFile
+
+          val file = if (junitFile.exists) junitFile else commandLineFile
+          val source = scala.io.Source.fromFile(file)
+          val lines = try source.mkString finally source.close()
+
+          val aspectDefinition = AspectDefinition("source", "source", Some(lines.parseJson.asJsObject))
+          param.asAdmin(Post("/v0/aspects", aspectDefinition)) ~> param.api(role).routes ~> check {
+            status shouldEqual StatusCodes.OK
+          }
+
+          val rightTagRightSource1 = Record("righttag-rightsource", "name", buildAspects("right"), Some("righttag"))
+
+          val all = List(rightTagRightSource1)
+
+          all.foreach(record => param.asAdmin(Post("/v0/records", record)) ~> param.api(role).routes ~> check {
+            status shouldEqual StatusCodes.OK
+          })
+
+          Get("/v0/records") ~> param.api(role).routes ~> check {
+            status shouldEqual StatusCodes.OK
+            val res = responseAs[RecordsPage[Record]]
+            res.records.length shouldEqual all.length
+          }
+
+          param.asAdmin(Delete("/v0/records?sourceTagToPreserve=righttag&sourceId=right")) ~> param.api(role).routes ~> check {
+            status shouldEqual StatusCodes.OK
+            responseAs[MultipleDeleteResult].count shouldEqual 0
+          }
+
+          Get("/v0/records") ~> param.api(role).routes ~> check {
+            status shouldEqual StatusCodes.OK
+            val res = responseAs[RecordsPage[Record]]
+            res.records.length shouldEqual 1
+            res.records.filter(record => record.id == ("righttag-rightsource")).length shouldEqual 1
+          }
+
+        }
+
         checkMustBeAdmin(role) {
           Delete("/v0/records?sourceTagToPreserve=blah&sourceId=blah2")
         }

